### PR TITLE
Provide meaningful rejection reasons

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Add meaningful rejection reasons for all gossip validators.

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java [hunk 1]
```diff
@@ -157,15 +161,19 @@ private InternalValidationResult singleAttestationChecks(final Attestation attes
...
 
               // The attestation's committee index (attestation.data.index) is for the correct
               // subnet.
               if (receivedOnSubnetId.isPresent()
                   && spec.computeSubnetForAttestation(state, attestation)
                       != receivedOnSubnetId.getAsInt()) {
-                return completedFuture(InternalValidationResult.REJECT);
+                return completedFuture(
+                    InternalValidationResult.reject(
+                        "Attestation received on incorrect subnet for specified committee index"));
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java]: add committee index in question?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java [hunk 2]
```diff
@@ -210,7 +224,8 @@ private InternalValidationResult singleAttestationChecks(final Attestation attes
                                 ancestorOfLMDVote ->
                                     ancestorOfLMDVote.equals(finalizedCheckpoint.getRoot()))
                             .orElse(false)) {
-                          return InternalValidationResult.REJECT;
+                          return InternalValidationResult.reject(
+                              "Attestation block root does not descent from finalized checkpoint");
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java:229]: below we use 'does not descend', do we want that same description here?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/InternalValidationResult.java [hunk 3]
```diff
@@ -48,6 +46,26 @@ public static InternalValidationResult create(
...
+
+  @FormatMethod
+  public static InternalValidationResult reject(
+      final String descriptionTemplate, final Object... args) {
+    return create(ValidationResultCode.REJECT, String.format(descriptionTemplate, args));
+  }
+
+  public static InternalValidationResult ignore(final String description) {
+    return create(ValidationResultCode.IGNORE, description);
+  }
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/InternalValidationResult.java]: with the other 2 as format methods, might be worth this one being a format so parameters can be passed consistently

**Review by rolfyone — APPROVED** [review]
Looks like a good iteration over all, some minor questions

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java]: Done.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java (hunk 2 above)
Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/AttestationValidator.java:229]: Done.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/InternalValidationResult.java (hunk 3 above)
Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/InternalValidationResult.java]: There's two versions, one that takes a single `String` and doesn't incur the cost of `String.format` and one that takes args and does call `String.format`.

Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/InternalValidationResult.java]: hmm, actually I can see the potential issue here, will remove the single string versions so we get compiler errors if you specify a `%s` but don't pass an actual value.

**Pull request merged**